src/parser/stage-2-ast.ts

```diff
@@ -678,9 +699,32 @@ function getName(
 }
 
 export function cstToAst(
-  cst: LiquidHtmlCST | ConcreteAttributeNode[],
+  cst: LiquidHtmlCST | LiquidCST | ConcreteAttributeNode[],
+  options: AstBuildOptions,
 ): LiquidHtmlNode[] {
   if (cst.length === 0) return [];
+
+  const builder = buildAst(cst, options);
+  const isStrictParser = options.mode === 'strict';
+
+  if (!isStrictParser || builder.cursor.length === 0) {
+    return builder.ast;
+  }
+
+  throw new LiquidHTMLASTParsingError(
+    `Attempting to end parsing before ${builder.parent?.type} '${getName(
+      builder.parent,
+    )}' was closed`,
+    builder.source,
+    builder.source.length - 1,
+    builder.source.length,
+  );
+}
```

Comment:
I'm of the opinion that the "default case"/"case you want" should be the flatest part of the code. I'd invert the conditions here. My logic being that this is what you'd expect while reading the code (if statements are reserved for exceptions). 

```
if (strictParser && builder.cursor.length !== 0) {
  throw err
}

return builder.ast;
```